### ISSUE TITLE
sparse-dense multiply output changes with dense matrix size

### ISSUE THREAD

**Issue by tmills** [cusplibrary/cusplibrary#73]
I have a sparse transition matrix (csr_matrix) A that I multiply by a dense array2d B into array2d C, where each column in B and C is an independent state vector. I'm hoping to optimize by increasing the number of columns but as B and C get wider the results in column 0 change. Here is a MWE:

```cuda
#include <stdio.h>
#include <iostream>

#include <cusp/csr_matrix.h>
#include <cusp/coo_matrix.h>
#include <cusp/print.h>
#include <cusp/functional.h>
#include <cusp/multiply.h>
#include <cusp/array2d.h>
#include <cusp/array1d.h>
#include <cusp/coo_matrix.h>
#include <cusp/print.h>
#include <cusp/gallery/poisson.h>

using namespace std;

int main()
{
  int state_size = 15;

  cusp::array2d<float, cusp::device_memory> pi(state_size, state_size);
  pi(1,0) = 0.19; pi(2,0) = 0.2; pi(3,0) = 0.21; pi(4,0) = 0.22; pi(5,0) = 0.18;
  pi(1,1) = 0.19; pi(2,2) = 0.2; pi(3,3) = 0.21; pi(4,4) = 0.22; pi(5,5) = 0.18;
  pi(1,6) = 0.19; pi(2,7) = 0.2; pi(3,8) = 0.21; pi(4,9) = 0.22; pi(5,10) = 0.18;
  pi(5,1) = 0.29; pi(6,2) = 0.3; pi(7,3) = 0.31; pi(8,4) = 0.32; pi(9,5) = 0.28;
  pi(5,6) = 0.29; pi(6,7) = 0.3; pi(7,8) = 0.31; pi(8,9) = 0.32; pi(9,10) = 0.28;
  
  cout << "Dense version of pi:" << endl;
  cusp::print(pi);
  cusp::csr_matrix<int,float,cusp::device_memory> pi_sparse(pi);
    
  for(int batch_size = 1; batch_size <= 6; batch_size++){
      cusp::array2d<float, cusp::device_memory> prev_mat(state_size, batch_size, 0.0f);
      cusp::array2d<float, cusp::device_memory> next_mat(state_size, batch_size, 0.0f);

      for(int i = 0; i < batch_size; i++){
          prev_mat(0, i) = 1;
      }
  
      cusp::multiply(pi_sparse, prev_mat, next_mat);
  
      cout << "Previous matrix:" << endl;
      cusp::print(prev_mat);
      cout << "Next matrix non-zeros:" << endl;
  
      for(int i = 0; i < batch_size; i++){
        for(int j = 0; j < state_size; j++){
             if(next_mat(j,i) > 0){
                 cout << "Value of next_mat[ " << j << ", " << i << "] = " << next_mat(j,i) << endl;
             }
          }
      }
  }
}
```

In the first pass where batch_size=1, the output is correct:
Value of next_mat[ 1, 0] = 0.19
Value of next_mat[ 2, 0] = 0.2
Value of next_mat[ 3, 0] = 0.21
Value of next_mat[ 4, 0] = 0.22
Value of next_mat[ 5, 0] = 0.18

(in fact all columns of next_mat should look like this).
As batch_size increases (here 3), it adds odd values into rows that should be 0 based on the transition matrix:
Value of next_mat[ 2, 0] = 0.19
Value of next_mat[ 4, 0] = 0.2
Value of next_mat[ 5, 0] = 0.21
Value of next_mat[ 9, 0] = 2.262e+16
Value of next_mat[ 14, 0] = 4.2e-17

or puts the right values into the wrong rows (batch_size=5):
Value of next_mat[ 2, 0] = 0.19
Value of next_mat[ 3, 0] = 0.2
Value of next_mat[ 4, 0] = 0.21
Value of next_mat[ 6, 0] = 0.22
Value of next_mat[ 7, 0] = 0.18

Any help will be appreciated.

**Comment by tmills** [comment]
It seems this happens when the number of columns of the matrix is not a power of 2. I think this is because of the cusp/system/cuda/detail/multiply/csr_block_spmv.h multiply function that uses powers of 2 as NUM_COLS if the input number of columns is <= power of 2. Probably this function should throw an exception if the number of columns does not fit into those bins.

**Comment by sdalton1** [comment]
Thanks for pointing out the bug. The sparse-dense multiply is still experimental and was used primarily for blocked eigensolvers where power of 2 column lengths were assumed. I will push some updates to generalize the implementation.

**Comment by tmills** [comment]
Is there any reason it's capped at 32? I just edited it myself to allow 64,128,256, etc. but at 64 I get the error: 

terminate called after throwing an instance of 'thrust::system::system_error'
  what():  cudaEventSynchronize in future::wait: an illegal memory access was encountered

I can look into it a bit at some point but if you know offhand that will save me some time.

**Comment by sdalton1** [comment]
There is no real limitation on 32 columns. You can extend the functionality by either looping over the columns and launching separate kernels but that causes the matrix to be loaded multiple times. Looping over the columns within a single launch should be a simple matter of adding 2 for loops within the existing kernel to accumulate values for say 64 columns in 2 32 column width passes. If you upload your code I can take a look to see how you are performing the updates now.

**Comment by tmills** [comment]
Yes I was able to fix on my code by splitting the batches into mini-batches of 32,16,8,4,2,1. But to get more performance I wanted to allow batches of 64 so I modified the code as:

```
if (x.num_cols <= 64) {
    __spmv_csr_block<64>(exec, A, x, y, initialize, combine, reduce);
    return;
}
```

and gave it a matrix with 64 columns and got the above-mentioned error. Probably there is something in the index calculations that breaks when num cols = 64 but I haven't had time to track it down yet.
